<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://neurobioinfo.github.io/Step3/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Step 4: Application of Ensemblux - Ensemblux</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Step 4: Application of Ensemblux";
        var mkdocs_page_input_path = "Step3.md";
        var mkdocs_page_url = "/Step3/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Ensemblux
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">The Ensemblux algorithm</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../overview/">Ensemblux algorithm overview</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">The Ensemblux Pipeline</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../overview_pipeline/">Ensemblux pipeline overview</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step0/">Step 1: Set up</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step1/">Step 2: Preparation of input files</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step2/">Step 3: Genetic demultiplexing by constituent tools</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Step 4: Application of Ensemblux</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#ensemblux-parameters">Ensemblux parameters</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#applying-the-ensemblux-algorithm">Applying the Ensemblux algorithm</a>
    </li>
    </ul>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../reference/">Execution parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../outputs/">Ensemblux outputs</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../midbrain_download/">Downloading data</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Dataset1/">Ensemblux with prior genotype information</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../contributing/">Help and Feedback</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Acknowledgement/">Acknowledgement</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../LICENSE/">License</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Ensemblux</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>The Ensemblux Pipeline &raquo;</li>
      <li class="breadcrumb-item active">Step 4: Application of Ensemblux</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="step-4-application-of-ensemblux">Step 4: Application of Ensemblux</h1>
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#ensemblux-parameters">Ensemblux parameters</a></li>
<li><a href="#applying-the-ensemblux-algorithm">Applying the Ensemblux algorithm</a></li>
</ul>
<hr />
<h2 id="introduction">Introduction</h2>
<p>In Step 4, we will process the output files from the constituent genetic demultiplexing tools with the Ensemblux framework. Ensemblux processes the output files in a three-step pipeline to identify the most probable sample label for each cell based on the predictions of the constituent tools:</p>
<p><strong>Step 1: Probabilistic-weighted ensemble</strong> <br />
In Step 1,  Ensemblux utilizes an unsupervised weighting model to identify the most probable sample label for each cell. Ensemblux weighs each constituent toolâ€™s assignment probability distribution by its estimated balanced accuracy for the dataset. The weighted assignment probabilities across all four constituent tools are then used to inform the most probable sample label for each cell.</p>
<p><strong>Step 2: Graph-based doublet detection</strong> <br />
In Step 2, Ensemblux utilizes a graph-based approach to identify doublets that were incorrectly labeled as singlets in Step 1. Pooled cells are embedded into PCA space and the most confident doublets in the pool (nCD) are identified. Then, based on the Euclidean distance in PCA space, the pooled cells that surpass the percentile threshold (pT) of the nearest neighbour frequency to the confident doublets are labelled as doublets by Ensemblux. Ensemblux performs an automated parameter sweep to identify the optimal nCD and pT values; however, user can opt to manually define these parameters. </p>
<p><strong>Step 3: Ensemble-independent doublet detection</strong> <br />
In Step 3, Ensemblux utilizes an ensemble-independent approach to further improve doublet detection. Here, cells that are labelled as doublets by Demuxalot or Vireo are labelled as doublets by Ensemblux; however, users can nominate different tools to utilize for Step 3, depending on the desired doublet detection stringency. </p>
<hr />
<h2 id="ensemblux-parameters">Ensemblux parameters</h2>
<p>Users can choose to run each step of the Ensemblux framework sequentially (Steps 1 to 3) or can opt to skip certain steps. While Step 1 is necessary to generate the ensemble sample labels, Steps 2 and 3 were implemented to improve Ensemblux's ability to identify doublets; thus, if users do not want to prioritize doublet detection, they may skip Steps 2 and/or 3. Nonetheless, we demonstrated in our <strong>pre-print manuscript</strong> that utilizing the entire Ensemblux framework is important for maximizing the demultiplexing accuracy. Users can define which steps of the Ensemblux framework they want to utilize in the adjustable parameters file.</p>
<p>The adjustable parameters file (<code>ensemblux_config.ini</code>) is located in <code>~/working_directory/job_info/configs/</code>. For a comprehensive description of how to adjust the analytical parameters of the Ensemblux pipeline please see <a href="../reference/">Execution parameters</a>. The following parameters are adjustable when applying the Ensemblux algorithm:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Default</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><ins><strong>Pool parameters</strong></ins></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_sample_size</td>
<td style="text-align: left;">NULL</td>
<td style="text-align: left;">Number of samples multiplexed in the pool.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_expected_doublet_rate</td>
<td style="text-align: left;">NULL</td>
<td style="text-align: left;">Expected doublet rate for the pool. If using 10X Genomics, the expected doublet rate can be estimated based on the number of recovered cells. For more information see <a href="https://kb.10xgenomics.com/hc/en-us/articles/360059124751-Why-is-the-multiplet-rate-different-for-the-Next-GEM-Single-Cell-3-LT-v3-1-assay-compared-to-other-single-cell-applications">10X Genomics Documentation</a>.</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Set up parameters</strong></ins></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_merge_constituents</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to merge the output files of the constituent demultiplexing tools. If running Ensemblux on a pool for the first time, this parameter should be set to "Yes". Subsequent runs of Ensemblux (e.g., parameter optimization) can have this parameter set to "No" as the pipeline will automatically detect the previously generated merged file.</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Step 1 parameters: Probabilistic-weighted ensemble</strong></ins></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_probabilistic_weighted_ensemble</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to perform Step 1: Probabilistic-weighted ensemble. If running Ensemblux on a pool for the first time, this parameter should be set to "Yes". Subsequent runs of Ensemblux (e.g., parameter optimization) can have this parameter set to "No" as the pipeline will automatically detect the previously generated Step 1 output file.</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Step 2 parameters: Graph-based doublet detection</strong></ins></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_preliminary_parameter_sweep</td>
<td style="text-align: left;">No</td>
<td style="text-align: left;">Whether or not to perform a preliminary parameter sweep for Step 2: Graph-based doublet detection. Users should utilize the preliminary parameter sweep if they wish to manually define the number of confident doublets in the pool (nCD) and the percentile threshold of the nearest neighour frequency (pT), which can be defined in the following two parameters, respectively.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_nCD</td>
<td style="text-align: left;">NULL</td>
<td style="text-align: left;">Manually defined number of confident doublets in the pool (nCD). Value can be informed by the output files generated by setting PAR_ensemblux_preliminary_parameter_sweep to "Yes".</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_pT</td>
<td style="text-align: left;">NULL</td>
<td style="text-align: left;">Manually defined percentile threshold of the nearest neighour frequency (pT). Value can be informed by the output files generated by setting PAR_ensemblux_preliminary_parameter_sweep to "Yes".</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_graph_based_doublet_detection</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to perform Step 2: Graph-based doublet detection. If PAR_ensemblux_nCD and PAR_ensemblux_pT are not defined by the user (NULL), Ensemblux will automatically determine the optimal parameter values using an unsupervised parameter sweep. If PAR_ensemblux_nCD and PAR_ensemblux_pT are defined by the user, graph-based doublet detection will be performed with the user-defined values.</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Step 3 parameters: Ensemble-independent doublet detection</strong></ins></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_preliminary_ensemble_independent_doublet</td>
<td style="text-align: left;">No</td>
<td style="text-align: left;">Whether or not to perform a preliminary parameter sweep for Step 3: Ensemble-independent doublet detection. Users should utilize the preliminary parameter sweep if they wish to manually define which constituent tools to utilize for ensemble-independent doublet detection. Users can define which tools to utilize for ensemble-independent doublet detection in the following parameters.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_ensemble_independent_doublet</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to perform Step 3: Ensemble-independent doublet detection.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_doublet_Demuxalot_threshold</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to label doublets identified by Demuxalot as doublets. Only doublets with assignment probabilities exceeding Demuxalot's recommended probability threshold will be labeled as doublets by Ensemblux.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_doublet_Demuxalot_no_threshold</td>
<td style="text-align: left;">No</td>
<td style="text-align: left;">Whether or not to label doublets identified by Demuxalot as doublets, regardless of the corresponding assignment probability.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_doublet_Demuxlet_threshold</td>
<td style="text-align: left;">No</td>
<td style="text-align: left;">Whether or not to label doublets identified by Demuxlet as doublets. Only doublets with assignment probabilities exceeding Demuxlet's recommended probability threshold will be labeled as doublets by Ensemblux.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_doublet_Demuxlet_no_threshold</td>
<td style="text-align: left;">No</td>
<td style="text-align: left;">Whether or not to label doublets identified by Demuxlet as doublets, regardless of the corresponding assignment probability.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_doublet_Souporcell_threshold</td>
<td style="text-align: left;">No</td>
<td style="text-align: left;">Whether or not to label doublets identified by Souporcell as doublets. Only doublets with assignment probabilities exceeding Souporcell's recommended probability threshold will be labeled as doublets by Ensemblux.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_doublet_Souporcell_no_threshold</td>
<td style="text-align: left;">No</td>
<td style="text-align: left;">Whether or not to label doublets identified by Souporcell as doublets, regardless of the corresponding assignment probability.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_doublet_Vireo_threshold</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to label doublets identified by Vireo as doublets. Only doublets with assignment probabilities exceeding Vireo's recommended probability threshold will be labeled as doublets by Ensemblux.</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_doublet_Vireo_no_threshold</td>
<td style="text-align: left;">No</td>
<td style="text-align: left;">Whether or not to label doublets identified by Vireo as doublets, regardless of the corresponding assignment probability.</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Confidence score parameters</strong></ins></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblux_compute_singlet_confidence</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to compute Ensemblux's singlet confidence score. This will define low confidence assignments which should be removed from downstream analyses.</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="applying-the-ensemblux-algorithm">Applying the Ensemblux algorithm</h2>
<p>To apply the Ensemblux algorithm use the following code:</p>
<pre><code>Ensemblux_HOME=/path/to/ensemblux.pip
Ensemblux_PWD=/path/to/working_directory

bash $Ensemblux_HOME/launch_ensemblux.sh -d $Ensemblux_PWD --step ensembluxing
</code></pre>
<p>If the Ensemblux algorithm completed successfully, the following files should be available in <code>~/working_directory/ensemblux</code></p>
<pre><code>working_directory
â””â”€â”€ ensemblux
    â”œâ”€â”€ confidence
    â”‚   â””â”€â”€ Ensemblux_final_cell_assignment.csv
    â”œâ”€â”€ constituent_tool_merge.csv
    â”œâ”€â”€ EID
    â”‚   â”œâ”€â”€ Doublet_overlap_no_threshold.pdf
    â”‚   â”œâ”€â”€ Doublet_overlap_threshold.pdf
    â”‚   â”œâ”€â”€ Number_Ensemblux_doublets_EID_no_threshold.pdf
    â”‚   â”œâ”€â”€ Number_Ensemblux_doublets_EID_threshold.pdf
    â”‚   â””â”€â”€ Step3_cell_assignment.csv
    â”œâ”€â”€ GBD
    â”‚   â”œâ”€â”€ optimal_nCD.pdf
    â”‚   â”œâ”€â”€ optimal_pT.pdf
    â”‚   â”œâ”€â”€ PC1_var_contrib.pdf
    â”‚   â”œâ”€â”€ PC2_var_contrib.pdf
    â”‚   â”œâ”€â”€ PCA1_graph_based_doublet_detection.pdf
    â”‚   â”œâ”€â”€ PCA2_graph_based_doublet_detection.pdf
    â”‚   â”œâ”€â”€ PCA3_graph_based_doublet_detection.pdf
    â”‚   â”œâ”€â”€ PCA_plot.pdf
    â”‚   â”œâ”€â”€ PCA_scree_plot.pdf
    â”‚   â””â”€â”€ Step2_cell_assignment.csv
    â””â”€â”€ PWE
        â”œâ”€â”€ ARI_demultiplexing_tools.pdf
        â”œâ”€â”€ BA_demultiplexing_tools.pdf
        â”œâ”€â”€ Balanced_accuracy_summary.csv
        â””â”€â”€ Step1_cell_assignment.csv
</code></pre>
<p>For a comprehensive description of the Ensemblux algorithm output files, please see <a href="../outputs/">Ensemblux outputs</a>.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../Step2/" class="btn btn-neutral float-left" title="Step 3: Genetic demultiplexing by constituent tools"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../reference/" class="btn btn-neutral float-right" title="Execution parameters">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../Step2/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../reference/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
