<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="description" content="None" /><link rel="canonical" href="https://neurobioinfo.github.io/" />
      <link rel="shortcut icon" href="img/favicon.ico" />
    <title>Ensemblex</title>
    <link rel="stylesheet" href="css/theme.css" />
    <link rel="stylesheet" href="css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Home";
        var mkdocs_page_input_path = "index.md";
        var mkdocs_page_url = "/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="." class="icon icon-home"> Ensemblex
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href=".">Home</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#contents">Contents</a>
    </li>
    </ul>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">The Ensemblex algorithm</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="overview/">Ensemblex algorithm overview</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">The Ensemblex Pipeline</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="overview_pipeline/">Ensemblex pipeline overview</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="installation/">Installation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Step0/">Step 1: Set up</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Step1/">Step 2: Preparation of input files</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Step2/">Step 3: Genetic demultiplexing by constituent tools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Step3/">Step 4: Application of Ensemblex</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="reference/">Execution parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="outputs/">Ensemblex outputs</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="midbrain_download/">Downloading data</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Dataset1/">Ensemblex with prior genotype information</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="contributing/">Help and Feedback</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Acknowledgement/">Acknowledgement</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="LICENSE/">License</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href=".">Ensemblex</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
      <li class="breadcrumb-item active">Home</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="welcome-to-the-ensemblex-documentation">Welcome to the Ensemblex documentation!</h1>
<p>Ensemblex is an accuracy-weighted ensemble framework for genetic demultiplexing of pooled single-cell RNA seqeuncing (scRNAseq) data. By addressing the limitiations of individual genetic demultiplexing tools, we demonstrated that Ensemblex:</p>
<ul>
<li>Achieves higher demultiplexing accuracy</li>
<li>Limits the introduction of technical noise into scRNAseq analysis </li>
<li>Retains a high proportion of cells for downstream analyses. </li>
</ul>
<p>The ensemble method capitalizes on the added confidence of combining distinct statistical frameworks for genetic demultiplexing, but the modular algorithm can adapt to the overall performance of its constituent tools on the respective dataset, making it resilient against a poorly performing constituent tool. </p>
<p>Ensemblex can be used to demultiplex pools <strong>with</strong> or <strong>without</strong> prior genotype information. When demultiplexing <strong>with</strong> prior genotype information, Ensemblex leverages the sample assignments of four individual, constituent genetic demultiplexing tools:</p>
<ol>
<li>Demuxalot (<a href="https://www.biorxiv.org/content/10.1101/2021.05.22.443646v2.abstract">Rogozhnikov et al. </a>)</li>
<li>Demuxlet (<a href="https://www.nature.com/articles/nbt.4042">Kang et al. </a>)</li>
<li>Souporcell (<a href="https://www.nature.com/articles/s41592-020-0820-1">Heaton et al. </a>)</li>
<li>Vireo-GT (<a href="https://link.springer.com/article/10.1186/s13059-019-1865-2">Huang et al. </a>)</li>
</ol>
<p>When demultiplexing <strong>without</strong> prior genotype information, Ensemblex leverages the sample assignments of four individual, constituent genetic demultiplexing tools:</p>
<ol>
<li>Demuxalot (<a href="https://www.biorxiv.org/content/10.1101/2021.05.22.443646v2.abstract">Rogozhnikov et al. </a>)</li>
<li>Freemuxlet (<a href="https://www.nature.com/articles/nbt.4042">Kang et al. </a>)</li>
<li>Souporcell (<a href="https://www.nature.com/articles/s41592-020-0820-1">Heaton et al. </a>)</li>
<li>Vireo (<a href="https://link.springer.com/article/10.1186/s13059-019-1865-2">Huang et al. </a>)</li>
</ol>
<p>Upon demultiplexing pools with each of the four constituent genetic demultiplexing tools, Ensemblex processes the output files in a three-step pipeline to identify the most probable sample label for each cell based on the predictions of the constituent tools:</p>
<p><strong>Step 1</strong>: Probabilistic-weighted ensemble <br />
<strong>Step 2</strong>: Graph-based doublet detection <br />
<strong>Step 3</strong>: Ensemble-independent doublet detection <br /></p>
<p>As output, Ensemblex returns its own cell-specific sample labels and corresponding assignment probabilities and singlet confidence score, as well as the sample labels and corresponding assignment probabilities for each of its constituents. The demultiplexed sample labels could then be used to perform downstream analyses.</p>
<p align="center">
 <img src="https://github.com/neurobioinfo/ensemblex/assets/97498007/d4161be8-8bc6-4c08-af60-4ee3f486e940" width="650" height="100">
 </p>

<p><strong>Figure 1. Overview of the Ensemblex worflow.</strong> <strong>A)</strong> The Ensemblex workflow begins with demultiplexing pooled samples by each of the constituent tools. The outputs from each individual demultiplexing tool are then used as input into the Ensemblex framework.  <strong>B)</strong> The Ensemblex framework comprises three distinct steps that are assembled into a pipeline: 1) accuracy-weighted probabilistic ensemble, 2) graph-based doublet detection, and 3) ensemble-independent doublet detection.  <strong>C)</strong> As output, Ensemblex returns its own sample-cell assignments as well as the sample-cell assignments of each of its constituent tools.  <strong>D)</strong> Ensemblex's sample-cell assignments can be used to perform downstream analysis on the pooled scRNAseq data. </p>
<p>To facilitate the application of Ensemblex, we provide a pipeline that demultiplexes pooled cells by each of the individual constituent genetic demultiplexing tools and processes the outputs with the Ensemblex algorithm. In this documentation, we outline each step of the Ensemblex pipeline, illustrate how to run the pipeline, define best practices, and provide a tutorial with pubicly available datasets. </p>
<p>For a comprehensive descripttion of Ensemblex, ground-truth benchmarking, and application to real-world datasets, see our pre-print manuscript: <strong>Pre-print</strong></p>
<hr />
<h2 id="contents">Contents</h2>
<ul>
<li>The Ensemblex Algorithm:<ul>
<li><a href="overview/">Ensemblex algorithm overview</a></li>
</ul>
</li>
<li>The Ensemblex Pipeline:<ul>
<li><a href="overview_pipeline/">Ensemblex pipeline overview</a></li>
<li><a href="installation/">Installation</a></li>
<li><a href="Step0/">Step 1: Set up</a></li>
<li><a href="Step1/">Step 2: Preparation of inpute files</a></li>
<li><a href="Step2/">Step 3: Genetic demultiplexing by constituent tools</a></li>
<li><a href="Step3/">Step 4: Application of Ensemblex</a></li>
</ul>
</li>
<li>Documentation:    <ul>
<li><a href="reference/">Execution parameters</a> </li>
<li><a href="outputs/">Ensemblex outputs</a> </li>
</ul>
</li>
<li>Tutorial:<ul>
<li><a href="midbrain_download/">Downloading data</a></li>
<li><a href="Dataset1/">Ensemblex with prior genotype information</a>            </li>
</ul>
</li>
<li>About:<ul>
<li><a href="contributing/">Help and Feedback</a></li>
<li><a href="Acknowledgement/">Acknowledgement</a></li>
<li><a href="LICENSE/">License</a></li>
</ul>
</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="overview/" class="btn btn-neutral float-right" title="Ensemblex algorithm overview">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
      <span><a href="overview/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="js/jquery-3.6.0.min.js"></script>
    <script>var base_url = ".";</script>
    <script src="js/theme_extra.js"></script>
    <script src="js/theme.js"></script>
      <script src="search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>

<!--
MkDocs version : 1.5.2
Build Date UTC : 2024-06-14 18:48:04.453186+00:00
-->
