<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://neurobioinfo.github.io/Step2/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Step 3: Genetic demultiplexing by constituent tools - Ensemblex</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Step 3: Genetic demultiplexing by constituent tools";
        var mkdocs_page_input_path = "Step2.md";
        var mkdocs_page_url = "/Step2/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Ensemblex
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">The Ensemblex algorithm</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../overview/">Ensemblex algorithm overview</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">The Ensemblex Pipeline</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../overview_pipeline/">Ensemblex pipeline overview</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step0/">Step 1: Set up</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step1/">Step 2: Preparation of input files</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Step 3: Genetic demultiplexing by constituent tools</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#demultiplexing-with-prior-genotype-information">Demultiplexing with prior genotype information</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#demuxalot">Demuxalot</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#demuxlet">Demuxlet</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#souporcell">Souporcell</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#vireo-gt">Vireo-GT</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#demultiplexing-without-prior-genotype-information">Demultiplexing without prior genotype information</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#freemuxlet">Freemuxlet</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#souporcell_1">Souporcell</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#vireo">Vireo</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#demuxalot_1">Demuxalot</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step3/">Step 4: Application of Ensemblex</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../reference/">Execution parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../outputs/">Ensemblex outputs</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../midbrain_download/">Downloading data</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Dataset1/">Ensemblex with prior genotype information</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../contributing/">Help and Feedback</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Acknowledgement/">Acknowledgement</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../LICENSE/">License</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Ensemblex</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>The Ensemblex Pipeline &raquo;</li>
      <li class="breadcrumb-item active">Step 3: Genetic demultiplexing by constituent tools</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="step-3-genetic-demultiplexing-by-constituent-demultiplexing-tools">Step 3: Genetic demultiplexing by constituent demultiplexing tools</h1>
<p>In Step 3, we will demultiplex the pooled samples with each of Ensemblex's constituent genetic demultiplexing tools. The constituent genetic demultiplexing tools will vary depending on the version of the Ensemblex pipeline being used:</p>
<ul>
<li><a href="#demultiplexing-with-prior-genotype-information">Demultiplexing with prior genotype information</a></li>
<li><a href="#demultiplexing-without-prior-genotype-information">Demultiplexing without prior genotype information</a></li>
</ul>
<p><strong>NOTE</strong>: The analytical parameters for each constiuent tool can be adjusted using the the <code>ensemblex_config.ini</code> file located in <code>~/working_directory/job_info/configs</code>. For a comprehensive description of how to adjust the analytical parameters of the Ensemblex pipeline please see <a href="../reference/">Execution parameters</a>. </p>
<hr />
<h2 id="demultiplexing-with-prior-genotype-information">Demultiplexing with prior genotype information</h2>
<p>When demultiplexing <strong>with</strong> prior genotype information, Ensemblex leverages the sample labels from</p>
<ul>
<li><a href="#demuxalot">Demuxalot</a></li>
<li><a href="#demuxlet">Demuxlet</a></li>
<li><a href="#souporcell">Souporcell</a></li>
<li><a href="#vireo-gt">Vireo-GT</a></li>
</ul>
<hr />
<h4 id="demuxalot">Demuxalot</h4>
<p>To run Demuxalot use the following code:</p>
<pre><code>ensemblex_HOME=/path/to/ensemblex.pip
ensemblex_PWD=/path/to/working_directory

bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step demuxalot
</code></pre>
<p>If Demuxalot completed successfully, the following files should be available in <code>~/working_directory/demuxalot</code></p>
<pre><code>working_directory
└── demuxalot
    ├── Demuxalot_result.csv
    └── new_snps_single_file.betas
</code></pre>
<hr />
<h4 id="demuxlet">Demuxlet</h4>
<p>To run Demuxlet use the following code:</p>
<pre><code>ensemblex_HOME=/path/to/ensemblex.pip
ensemblex_PWD=/path/to/working_directory

bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step demuxlet
</code></pre>
<p>If Demuxlet completed successfully, the following files should be available in <code>~/working_directory/demuxlet</code></p>
<pre><code>working_directory
└── demuxlet
    ├── outs.best
    ├── pileup.cel.gz
    ├── pileup.plp.gz
    ├── pileup.umi.gz
    └── pileup.var.gz
</code></pre>
<hr />
<h4 id="souporcell">Souporcell</h4>
<p>To run Souporcell use the following code:</p>
<pre><code>ensemblex_HOME=/path/to/ensemblex.pip
ensemblex_PWD=/path/to/working_directory

bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step souporcell
</code></pre>
<p>If Souporcell completed successfully, the following files should be available in <code>~/working_directory/souporcell</code></p>
<pre><code>working_directory
└── souporcell
    ├── alt.mtx
    ├── cluster_genotypes.vcf
    ├── clusters_tmp.tsv
    ├── clusters.tsv
    ├── fq.fq
    ├── minimap.sam
    ├── minitagged.bam
    ├── minitagged_sorted.bam
    ├── minitagged_sorted.bam.bai
    ├── Pool.vcf
    ├── ref.mtx
    └── soup.txt
</code></pre>
<hr />
<h4 id="vireo-gt">Vireo-GT</h4>
<p>To run Vireo-GT use the following code:</p>
<pre><code>ensemblex_HOME=/path/to/ensemblex.pip
ensemblex_PWD=/path/to/working_directory

bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step vireo
</code></pre>
<p>If Vireo-GT completed successfully, the following files should be available in <code>~/working_directory/vireo_gt</code></p>
<pre><code>working_directory
└── vireo_gt
    ├── cellSNP.base.vcf.gz
    ├── cellSNP.cells.vcf.gz
    ├── cellSNP.samples.tsv
    ├── cellSNP.tag.AD.mtx
    ├── cellSNP.tag.DP.mtx
    ├── cellSNP.tag.OTH.mtx
    ├── donor_ids.tsv
    ├── fig_GT_distance_estimated.pdf
    ├── fig_GT_distance_input.pdf
    ├── GT_donors.vireo.vcf.gz
    ├── _log.txt
    ├── prob_doublet.tsv.gz
    ├── prob_singlet.tsv.gz
    └── summary.tsv
</code></pre>
<hr />
<p>Upon demultiplexing the pooled samples with each of Ensemblex's constituent genetic demultiplexing tools, we can proceed to Step 4 where we will process the output files of the consituent tools with the Ensemblex algorithm to generate the ensemble sample classifications: <a href="../Step3/">Application of Ensemblex</a></p>
<hr />
<h2 id="demultiplexing-without-prior-genotype-information">Demultiplexing without prior genotype information</h2>
<p>When demultiplexing <strong>without</strong> prior genotype information, Ensemblex leverages the sample labels from</p>
<ul>
<li><a href="#freemuxlet">Freemuxlet</a></li>
<li><a href="#souporcell">Souporcell</a></li>
<li><a href="#vireo">Vireo</a></li>
<li><a href="#demuxalot">Demuxalot</a></li>
</ul>
<hr />
<h4 id="freemuxlet">Freemuxlet</h4>
<p>To run Freemuxlet use the following code:</p>
<pre><code>ensemblex_HOME=/path/to/ensemblex.pip
ensemblex_PWD=/path/to/working_directory

bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step freemuxlet
</code></pre>
<p>If Freemuxlet completed successfully, the following files should be available in <code>~/working_directory/freemuxlet</code></p>
<pre><code>working_directory
└── freemuxlet
    ├── outs.clust1.samples.gz
    ├── outs.clust1.vcf
    ├── outs.lmix
    ├── pileup.cel.gz
    ├── pileup.plp.gz
    ├── pileup.umi.gz
    └── pileup.var.gz
</code></pre>
<hr />
<h4 id="souporcell_1">Souporcell</h4>
<p>To run Souporcell use the following code:</p>
<pre><code>ensemblex_HOME=/path/to/ensemblex.pip
ensemblex_PWD=/path/to/working_directory

bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step souporcell
</code></pre>
<p>If Souporcell completed successfully, the following files should be available in <code>~/working_directory/souporcell</code></p>
<pre><code>working_directory
└── souporcell
    ├── alt.mtx
    ├── cluster_genotypes.vcf
    ├── clusters_tmp.tsv
    ├── clusters.tsv
    ├── fq.fq
    ├── minimap.sam
    ├── minitagged.bam
    ├── minitagged_sorted.bam
    ├── minitagged_sorted.bam.bai
    ├── Pool.vcf
    ├── ref.mtx
    └── soup.txt
</code></pre>
<hr />
<h4 id="vireo">Vireo</h4>
<p>To run Vireo use the following code:</p>
<pre><code>ensemblex_HOME=/path/to/ensemblex.pip
ensemblex_PWD=/path/to/working_directory

bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step vireo
</code></pre>
<p>If Vireo completed successfully, the following files should be available in <code>~/working_directory/vireo</code></p>
<pre><code>working_directory
└── vireo
    ├── cellSNP.base.vcf.gz
    ├── cellSNP.cells.vcf.gz
    ├── cellSNP.samples.tsv
    ├── cellSNP.tag.AD.mtx
    ├── cellSNP.tag.DP.mtx
    ├── cellSNP.tag.OTH.mtx
    ├── donor_ids.tsv
    ├── fig_GT_distance_estimated.pdf
    ├── GT_donors.vireo.vcf.gz
    ├── _log.txt
    ├── prob_doublet.tsv.gz
    ├── prob_singlet.tsv.gz
    └── summary.tsv
</code></pre>
<hr />
<h4 id="demuxalot_1">Demuxalot</h4>
<p><strong>NOTE</strong>: Because the Demuxalot algorithm requires prior genotype information, the Ensemblex pipeline uses the predicted vcf file generated by Freemuxlet as input into Demuxalot when prior genotype information is not available. Therefore, it is important to wait for Freemuxlet to complete before running Demuxalot. To check if the required Freemuxlet-generated vcf file is available prior to running Demuxalot, you can use the following code:</p>
<pre><code>if test -f /path/to/working_directory/freemuxlet/outs.clust1.vcf; then
  echo &quot;File exists.&quot;
fi
</code></pre>
<p>Upon confirming that the required Freemuxlet-generated file exists, we can run Demuxalot using the following code:</p>
<pre><code>ensemblex_HOME=/path/to/ensemblex.pip
ensemblex_PWD=/path/to/working_directory

bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step demuxalot
</code></pre>
<p>If Demuxalot completed successfully, the following files should be available in <code>~/working_directory/demuxalot</code></p>
<pre><code>working_directory
└── demuxalot
    ├── Demuxalot_result.csv
    └── new_snps_single_file.betas
</code></pre>
<hr />
<p>Upon demultiplexing the pooled samples with each of Ensemblex's constituent genetic demultiplexing tools, we can proceed to Step 4 where we will process the output files of the consituent tools with the Ensemblex algorithm to generate the ensemble sample classifications: <a href="../Step3/">Application of Ensemblex</a></p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../Step1/" class="btn btn-neutral float-left" title="Step 2: Preparation of input files"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../Step3/" class="btn btn-neutral float-right" title="Step 4: Application of Ensemblex">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../Step1/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../Step3/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
