<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://neurobioinfo.github.io/Dataset1/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Ensemblex with prior genotype information - Ensemblex</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Ensemblex with prior genotype information";
        var mkdocs_page_input_path = "Dataset1.md";
        var mkdocs_page_url = "/Dataset1/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Ensemblex
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">The Ensemblex algorithm</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../overview/">Ensemblex algorithm overview</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">The Ensemblex Pipeline</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../overview_pipeline/">Ensemblex pipeline overview</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step0/">Step 1: Set up</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step1/">Step 2: Preparation of input files</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step2/">Step 3: Genetic demultiplexing by constituent tools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step3/">Step 4: Application of Ensemblex</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../reference/">Execution parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../outputs/">Ensemblex outputs</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../midbrain_download/">Downloading data</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Ensemblex with prior genotype information</a>
    <ul class="current">
    </ul>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../contributing/">Help and Feedback</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Acknowledgement/">Acknowledgement</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../LICENSE/">License</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Ensemblex</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Tutorial &raquo;</li>
      <li class="breadcrumb-item active">Ensemblex with prior genotype information</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="ensemblex-pipeline-with-prior-genotype-information">Ensemblex pipeline with prior genotype information</h2>
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#step-1-set-up">Step 1: Set up</a></li>
<li><a href="#step-2-preparation-of-input-files">Step 2: Preparation of input files</a></li>
<li><a href="#step-3-genetic-demultiplexing-by-constituent-tools">Step 3: Genetic demultiplexing by constituent tools</a></li>
<li><a href="#step-4-application-of-ensemblex">Step 4: Application of Ensemblex</a></li>
<li><a href="#resource-requirements">Resource requirements</a></li>
</ul>
<hr />
<h2 id="introduction">Introduction</h2>
<p>This guide illustrates how to use the Ensemblex pipeline to demultiplexed pooled scRNAseq samples with prior genotype information. Here, we will leverage a pooled scRNAseq dataset produced by <a href="https://www.nature.com/articles/s41588-021-00801-6">Jerber et al.</a>. This pool contains induced pluripotent cell lines (iPSC) from 9 healthy controls that were differentiated towards a dopaminergic neuron state. The Ensemblex pipeline is illustrated in the diagram below:</p>
<p align="center">
 <img src="https://github.com/neurobioinfo/ensemblex/assets/97498007/58dac8ce-c83a-44ef-aa0e-9c31e771388f" width="200" height="100">
 </p>

<p><strong>NOTE</strong>: To download the necessary files for the tutorial please see the <a href="../midbrain_download/">Downloading data</a> section of the Ensemblex documentation.</p>
<hr />
<h2 id="installation">Installation</h2>
<p>[to be completed]</p>
<p>module load StdEnv/2023
module load apptainer/1.2.4 </p>
<hr />
<h2 id="step-1-set-up">Step 1: Set up</h2>
<p>In Step 1, we will set up the working directory for the Ensemblex pipeline and decide which version of the pipeline we want to use.</p>
<p>First, create a dedicated folder for the analysis (hereafter referred to as the working directory). Then, define the path to the working directory and the path to ensemblex.pip:</p>
<pre><code>## Create and navigate to the working directory
cd ensemblex_tutorial
mkdir working_directory
cd ~/ensemblex_tutorial/working_directory

## Define the path to ensemblex.pip
ensemblex_HOME=~/ensemblex.pip

## Define the path to the working directory
ensemblex_PWD=~/ensemblex_tutorial/working_directory
</code></pre>
<p>Next, we can set up the working directory and choose the Ensemblex pipeline for demultiplexing with prior genotype information (<code>--step init-GT</code>) using the following code:</p>
<pre><code>bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step init-GT
</code></pre>
<p>After running the above code, the working directory should have the following structure:</p>
<pre><code>ensemblex_tutorial
└── working_directory
    ├── demuxalot
    ├── demuxlet
    ├── ensemblex_gt
    ├── input_files
    ├── job_info
    │   ├── configs
    │   │   └── ensemblex_config.ini
    │   ├── logs
    │   └── summary_report.txt
    ├── souporcell
    └── vireo_gt
</code></pre>
<p>Upon setting up the Ensemblex pipeline, we can proceed to Step 2 where we will prepare the input files for Ensemblex's constituent genetic demultiplexing tools.</p>
<hr />
<h2 id="step-2-preparation-of-input-files">Step 2: Preparation of input files</h2>
<p>In Step 2, we will define the necessary files needed for ensemblex's constituent genetic demultiplexing tools and will place them within the working directory. </p>
<p><strong>Note</strong>: For the tutorial we will be using the data downloaded in the <a href="../midbrain_download/">Downloading data</a> section of the Ensemblex documentation. </p>
<p>First, define all of the required files:</p>
<pre><code>BAM=~/ensemblex_tutorial/CellRanger/outs/possorted_genome_bam.bam

BAM_INDEX=~/ensemblex_tutorial/CellRanger/outs/possorted_genome_bam.bam.bai

BARCODES=~/ensemblex_tutorial/CellRanger/outs/filtered_gene_bc_matrices/refdata-cellranger-GRCh37/barcodes.tsv

SAMPLE_VCF=~/ensemblex_tutorial/sample_genotype/sample_genotype_merge.vcf

REFERENCE_VCF=~/ensemblex_tutorial/reference_files/common_SNPs_only.recode.vcf

REFERENCE_FASTA=~/ensemblex_tutorial/reference_files/genome.fa

REFERENCE_FASTA_INDEX=~/ensemblex_tutorial/reference_files/genome.fa.fai
</code></pre>
<p>Next, we will sort the pooled samples and reference .vcf files according to the .bam file and place them within the working directory:</p>
<pre><code>## Sort pooled samples .vcf file
bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD/input_files/pooled_samples.vcf --step sort --vcf $SAMPLE_VCF --bam $ensemblex_PWD/input_files/pooled_bam.bam

## Sort reference .vcf file
bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD/input_files/reference.vcf --step sort --vcf $SAMPLE_VCF --bam $ensemblex_PWD/input_files/pooled_bam.bam
</code></pre>
<p><strong>NOTE</strong>: To sort the vcf files we use the pipeline produced by the authors of Demuxlet/Freemuxlet (<a href="https://www.nature.com/articles/nbt.4042">Kang et al. </a>).</p>
<p>Next, we will place the remaining necessary files within the working directory:</p>
<pre><code>cp $BAM $ensemblex_PWD/input_files/pooled_bam.bam
cp $BAM_INDEX $ensemblex_PWD/input_files/pooled_bam.bam.bai 
cp $BARCODES $ensemblex_PWD/input_files/pooled_barcodes.tsv
cp $REFERENCE_FASTA $ensemblex_PWD/input_files/reference.fa
cp $REFERENCE_FASTA_INDEX $ensemblex_PWD/input_files/reference.fa.fai
</code></pre>
<p>After running the above code,  <code>$ensemblex_PWD/input_files</code> should contain the following files:</p>
<pre><code>input_files
├── pooled_bam.bam
├── pooled_bam.bam.bai
├── pooled_barcodes.tsv
├── pooled_samples.vcf
├── reference.fa
├── reference.fa.fai
└── reference.vcf
</code></pre>
<p><strong>NOTE</strong>: It is important that the file names match those listed above as they are necessary for the Ensemblex pipeline to recognize them.</p>
<hr />
<h2 id="step-3-genetic-demultiplexing-by-constituent-tools">Step 3: Genetic demultiplexing by constituent tools</h2>
<p>In Step 3, we will demultiplex the pooled samples with each of Ensemblex's constituent genetic demultiplexing tools:</p>
<ul>
<li><a href="#demuxalot">Demuxalot</a></li>
<li><a href="#demuxlet">Demuxlet</a></li>
<li><a href="#souporcell">Souporcell</a></li>
<li><a href="#vireo-gt">Vireo-GT</a></li>
</ul>
<p>First, we will navigate to the <code>ensemblex_config.ini</code> file to adjust the demultiplexing parameters for each of the constituent genetic demultiplexing tools:</p>
<pre><code>## Navigate to the .ini file
cd $ensemblex_PWD/job_info/configs

## Open the .ini file and adjust parameters directly in the terminal
nano ensemblex_config.ini
</code></pre>
<p>For the tutorial, we set the following parameters for the constituent genetic demultiplexing tools:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">PAR_demuxalot_genotype_names</td>
<td style="text-align: left;">'HPSI0115i-hecn_6,HPSI0214i-pelm_3,HPSI0314i-sojd_3,HPSI0414i-sebn_3,HPSI0514i-uenn_3,HPSI0714i-pipw_4,HPSI0715i-meue_5,HPSI0914i-vaka_5,HPSI1014i-quls_2'</td>
</tr>
<tr>
<td style="text-align: left;">PAR_demuxalot_prior_strength</td>
<td style="text-align: left;">100</td>
</tr>
<tr>
<td style="text-align: left;">PAR_demuxalot_minimum_coverage</td>
<td style="text-align: left;">200</td>
</tr>
<tr>
<td style="text-align: left;">PAR_demuxalot_minimum_alternative_coverage</td>
<td style="text-align: left;">10</td>
</tr>
<tr>
<td style="text-align: left;">PAR_demuxalot_n_best_snps_per_donor</td>
<td style="text-align: left;">100</td>
</tr>
<tr>
<td style="text-align: left;">PAR_demuxalot_genotypes_prior_strength</td>
<td style="text-align: left;">1</td>
</tr>
<tr>
<td style="text-align: left;">PAR_demuxalot_doublet_prior</td>
<td style="text-align: left;">0.25</td>
</tr>
<tr>
<td style="text-align: left;">PAR_demuxlet_field</td>
<td style="text-align: left;">GT</td>
</tr>
<tr>
<td style="text-align: left;">PAR_vireo_N</td>
<td style="text-align: left;">9</td>
</tr>
<tr>
<td style="text-align: left;">PAR_vireo_type</td>
<td style="text-align: left;">GT</td>
</tr>
<tr>
<td style="text-align: left;">PAR_vireo_processes</td>
<td style="text-align: left;">20</td>
</tr>
<tr>
<td style="text-align: left;">PAR_vireo_minMAF</td>
<td style="text-align: left;">0.1</td>
</tr>
<tr>
<td style="text-align: left;">PAR_vireo_minCOUNT</td>
<td style="text-align: left;">20</td>
</tr>
<tr>
<td style="text-align: left;">PAR_vireo_forcelearnGT</td>
<td style="text-align: left;">T</td>
</tr>
<tr>
<td style="text-align: left;">PAR_minimap2</td>
<td style="text-align: left;">'-ax splice -t 8 -G50k -k 21 -w 11 --sr -A2 -B8 -O12,32 -E2,1 -r200 -p.5 -N20 -f1000,5000 -n2 -m20 -s40 -g2000 -2K50m --secondary=no'</td>
</tr>
<tr>
<td style="text-align: left;">PAR_freebayes</td>
<td style="text-align: left;">'-iXu -C 2 -q 20 -n 3 -E 1 -m 30 --min-coverage 6'</td>
</tr>
<tr>
<td style="text-align: left;">PAR_vartrix_umi</td>
<td style="text-align: left;">TRUE</td>
</tr>
<tr>
<td style="text-align: left;">PAR_vartrix_mapq</td>
<td style="text-align: left;">30</td>
</tr>
<tr>
<td style="text-align: left;">PAR_vartrix_threads</td>
<td style="text-align: left;">8</td>
</tr>
<tr>
<td style="text-align: left;">PAR_souporcell_k</td>
<td style="text-align: left;">9</td>
</tr>
<tr>
<td style="text-align: left;">PAR_souporcell_t</td>
<td style="text-align: left;">8</td>
</tr>
</tbody>
</table>
<p>Now that the parameters have been defined, we can demultiplex the pools with the constituent genetic demultiplexing tools.</p>
<hr />
<h4 id="demuxalot">Demuxalot</h4>
<p>To run Demuxalot use the following code:</p>
<pre><code>bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step demuxalot
</code></pre>
<p>If Demuxalot completed successfully, the following files should be available in <code>$ensemblex_PWD/demuxalot</code>:</p>
<pre><code>demuxalot
    ├── Demuxalot_result.csv
    └── new_snps_single_file.betas
</code></pre>
<hr />
<h4 id="demuxlet">Demuxlet</h4>
<p>To run Demuxlet use the following code:</p>
<pre><code>bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step demuxlet
</code></pre>
<p>If Demuxlet completed successfully, the following files should be available in <code>$ensemblex_PWD/demuxlet</code>:</p>
<pre><code>demuxlet
    ├── outs.best
    ├── pileup.cel.gz
    ├── pileup.plp.gz
    ├── pileup.umi.gz
    └── pileup.var.gz
</code></pre>
<hr />
<h4 id="souporcell">Souporcell</h4>
<p>To run Souporcell use the following code:</p>
<pre><code>bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step souporcell
</code></pre>
<p>If Souporcell completed successfully, the following files should be available in <code>$ensemblex_PWD/souporcell</code>:</p>
<pre><code>souporcell
    ├── alt.mtx
    ├── cluster_genotypes.vcf
    ├── clusters_tmp.tsv
    ├── clusters.tsv
    ├── fq.fq
    ├── minimap.sam
    ├── minitagged.bam
    ├── minitagged_sorted.bam
    ├── minitagged_sorted.bam.bai
    ├── Pool.vcf
    ├── ref.mtx
    └── soup.txt
</code></pre>
<hr />
<h4 id="vireo">Vireo</h4>
<p>To run Vireo-GT use the following code:</p>
<pre><code>bash $ensemblex_HOME/launch_ensemblex.sh -d $ensemblex_PWD --step vireo
</code></pre>
<p>If Vireo-GT completed successfully, the following files should be available in <code>$ensemblex_PWD/vireo_gt</code>:</p>
<pre><code>vireo_gt
    ├── cellSNP.base.vcf.gz
    ├── cellSNP.cells.vcf.gz
    ├── cellSNP.samples.tsv
    ├── cellSNP.tag.AD.mtx
    ├── cellSNP.tag.DP.mtx
    ├── cellSNP.tag.OTH.mtx
    ├── donor_ids.tsv
    ├── fig_GT_distance_estimated.pdf
    ├── fig_GT_distance_input.pdf
    ├── GT_donors.vireo.vcf.gz
    ├── _log.txt
    ├── prob_doublet.tsv.gz
    ├── prob_singlet.tsv.gz
    └── summary.tsv
</code></pre>
<hr />
<p>Upon demultiplexing the pooled samples with each of Ensemblex's constituent genetic demultiplexing tools, we can proceed to Step 4 where we will process the output files of the consituent tools with the Ensemblex algorithm to generate the ensemble sample classifications</p>
<p><strong>NOTE</strong>: To minimize computation time for the tutorial, we have provided the necessary outpu files from the constituent tools <a href="https://github.com/neurobioinfo/ensemblex/tree/main/tutorial">here</a>. To access the files and place them in the working directory, use the following code:</p>
<pre><code>## Demuxalot
cd $ensemblex_PWD/demuxalot
wget https://github.com/neurobioinfo/ensemblex/blob/caad8c250566bfa9a6d7a78b77d2cc338468a58e/tutorial/Demuxalot_result.csv

## Demuxlet
cd $ensemblex_PWD/demuxlet
wget https://github.com/neurobioinfo/ensemblex/blob/caad8c250566bfa9a6d7a78b77d2cc338468a58e/tutorial/outs.best

## Souporcell
cd $ensemblex_PWD/souporcell
wget https://github.com/neurobioinfo/ensemblex/blob/caad8c250566bfa9a6d7a78b77d2cc338468a58e/tutorial/clusters.tsv

## Vireo
cd $ensemblex_PWD/vireo_gt
wget https://github.com/neurobioinfo/ensemblex/blob/caad8c250566bfa9a6d7a78b77d2cc338468a58e/tutorial/donor_ids.tsv

</code></pre>
<hr />
<h2 id="step-4-application-of-ensemblex">Step 4: Application of Ensemblex</h2>
<p>In Step 4, we will process the output files of the four constituent genetic demultiplexing tools with the three-step Ensemblex algorithm:</p>
<ul>
<li>Step 1: Probabilistic-weighted ensemble</li>
<li>Step 2: Graph-based doublet detection</li>
<li>Step 3: Step 3: Ensemble-independent doublet detection</li>
</ul>
<p>First, we will navigate to the <code>ensemblex_config.ini</code> file to adjust the demultiplexing parameters for the Ensemblex algorithm:</p>
<pre><code>## Navigate to the .ini file
cd $ensemblex_PWD/job_info/configs

## Open the .ini file and adjust parameters directly in the terminal
nano ensemblex_config.ini
</code></pre>
<p>For the tutorial, we set the following parameters for the Ensemblex algorithm:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><ins><strong>Pool parameters</strong></ins></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_sample_size</td>
<td style="text-align: left;">9</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_expected_doublet_rate</td>
<td style="text-align: left;">0.10</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Set up parameters</strong></ins></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_merge_constituents</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Step 1 parameters: Probabilistic-weighted ensemble</strong></ins></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_probabilistic_weighted_ensemble</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Step 2 parameters: Graph-based doublet detection</strong></ins></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_preliminary_parameter_sweep</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_nCD</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_pT</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_graph_based_doublet_detection</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Step 3 parameters: Ensemble-independent doublet detection</strong></ins></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_preliminary_ensemble_independent_doublet</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_ensemble_independent_doublet</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_doublet_Demuxalot_threshold</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_doublet_Demuxalot_no_threshold</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_doublet_Demuxlet_threshold</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_doublet_Demuxlet_no_threshold</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_doublet_Souporcell_threshold</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_doublet_Souporcell_no_threshold</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_doublet_Vireo_threshold</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_doublet_Vireo_no_threshold</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;"><ins><strong>Confidence score parameters</strong></ins></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">PAR_ensemblex_compute_singlet_confidence</td>
<td style="text-align: left;">Yes</td>
</tr>
</tbody>
</table>
<p>If Ensemblex completed successfully, the following files should be available in <code>$ensemblex_PWD/ensemblex_gt</code>:</p>
<pre><code>ensemblex_gt
├── confidence
│   └── ensemblex_final_cell_assignment.csv
├── constituent_tool_merge.csv
├── step1
│   ├── ARI_demultiplexing_tools.pdf
│   ├── BA_demultiplexing_tools.pdf
│   ├── Balanced_accuracy_summary.csv
│   └── step1_cell_assignment.csv
├── step2
│   ├── optimal_nCD.pdf
│   ├── optimal_pT.pdf
│   ├── PC1_var_contrib.pdf
│   ├── PC2_var_contrib.pdf
│   ├── PCA1_graph_based_doublet_detection.pdf
│   ├── PCA2_graph_based_doublet_detection.pdf
│   ├── PCA3_graph_based_doublet_detection.pdf
│   ├── PCA_plot.pdf
│   ├── PCA_scree_plot.pdf
│   └── Step2_cell_assignment.csv
└── step3
    ├── Doublet_overlap_no_threshold.pdf
    ├── Doublet_overlap_threshold.pdf
    ├── Number_ensemblex_doublets_EID_no_threshold.pdf
    ├── Number_ensemblex_doublets_EID_threshold.pdf
    └── Step3_cell_assignment.csv
</code></pre>
<p>Ensemblex's final assignments are described in the <code>ensemblex_final_cell_assignment.csv</code> file. </p>
<p>Specifically, the <code>ensemblex_assignment</code> column describes Ensemblex's final assignments after application of the singlet confidence threshold (i.e., singlets that fail to meet a singlet confidence of 1.0 are labelled as unassigned); we recomment that users use this column to label their cells for downstream analyses. The <code>ensemblex_best_assignment</code> column describes Ensemblex's best assignments, independent of the singlets confidence threshold (i.e., singlets that fail to meet a singlet confidence of 1.0 are <strong>NOT</strong> labelled as unassigned).</p>
<p>The cell barcodes listed under the <code>barcode</code> column can be used to add the <code>ensemblex_final_cell_assignment.csv</code> information to the metadata of a Seurat object. </p>
<hr />
<h2 id="resource-requirements">Resource requirements</h2>
<p>The following table describes the computational resources used in this tutorial for genetic demultiplexing by the constituent tools and application of the Ensemblex algorithm.</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Tool</th>
<th style="text-align: left;">Time</th>
<th style="text-align: left;">CPU</th>
<th style="text-align: left;">Memory</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">Demuxalot</td>
<td style="text-align: left;">01:34:59</td>
<td style="text-align: left;">6</td>
<td style="text-align: left;">12.95 GB</td>
</tr>
<tr>
<td style="text-align: left;">Demuxlet</td>
<td style="text-align: left;">03:16:03</td>
<td style="text-align: left;">6</td>
<td style="text-align: left;">138.32 GB</td>
</tr>
<tr>
<td style="text-align: left;">Souporcell</td>
<td style="text-align: left;">2-14:49:21</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">21.83 GB</td>
</tr>
<tr>
<td style="text-align: left;">Vireo</td>
<td style="text-align: left;">2-01:30:24</td>
<td style="text-align: left;">6</td>
<td style="text-align: left;">29.42 GB</td>
</tr>
<tr>
<td style="text-align: left;">Ensemblex</td>
<td style="text-align: left;">02:05:27</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">5.67 GB</td>
</tr>
</tbody>
</table>
<hr />
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../midbrain_download/" class="btn btn-neutral float-left" title="Downloading data"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../contributing/" class="btn btn-neutral float-right" title="Help and Feedback">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../midbrain_download/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../contributing/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
